Editor: Allow title to split onto multiple lines

aduth opened the pull request:
~~Blocked by #4824~~
Closes #1059
Fixes #1679

This pull request seeks to allow the editor post title to split onto multiple lines, rather than obscuring it with a long-content fade when the input loses focus.

It also introduces a new `<TextareaAutosize />` component to serve as a drop-in replacement for `<textarea />` in order to facilitate in this use-case.

Before|After
---|---
![before](https://cloud.githubusercontent.com/assets/1779930/14613941/cc785cc8-056c-11e6-9540-9dd893afc253.png)|![after](https://cloud.githubusercontent.com/assets/1779930/14613924/ba7623a2-056c-11e6-933f-90d643308710.png)

__Implementation notes:__

While the implementation does use [`autosize`](https://github.com/jackmoore/autosize), a third-party library, note that this is already used in the editor for the HTML editing tab, so the change in editor bundle size should be negligible if anything. A follow-up task may be to consider whether the HTML tab textarea can be updated to use this component instead.

__Testing instructions:__

Verify in a large and small viewport size that the editor title resizes to accommodate multiple lines as needed. Ensure that existing behavior remains unaffected (notably, focusing the input for a new post, including when navigating from an existing post to a new post).

1. Navigate to the [Reader](http://calypso.localhost:3000/post)
2. Start a new post using the New Post masterbar button
3. Note that the title field is auto-focused for your new post
4. Enter text into the title area
5. Note that the textarea expands to allow multiple line entry

mtias commented on client/post-editor/editor-title/index.jsx:
```diff
@@ -118,17 +97,17 @@ const EditorTitle = React.createClass( {
 						aria-label={ this.translate( 'Edit title' ) }
-						ref="titleInput" />
+						ref="titleInput"
+						rows="1" />
```
I assume this is the default?

aduth replied:
Nope, the `rows` default is `2` ([reference](https://docs.webplatform.org/wiki/html/elements/textarea))

mtias replied:
Right, I mean the default initial value for the component.

aduth replied:
>Right, I mean the default initial value for the component.

Still not certain I'm following, but an empty field will default to `1` rows, and grow as necessary via `autosize`. Or are you suggesting we use `rows="1"` as a default prop for `<TextareaAutosize />` ?

mtias replied:
I didn't know what the prop meant just by looking at it :)

mtias commented:
This is great, thanks for looking at it after I complained :)

It works well for me, and I wouldn't mind merging before #4824 if that may take longer.

mtias commented on client/post-editor/editor-title/style.scss:
```diff
@@ -1,10 +1,6 @@
 .editor-title {
```
Is there anything we could simplify markup-wise here? Some of these nested elements were needed for correctly rendering the fade effect, which we don't need anymore.

aduth replied:
>Is there anything we could simplify markup-wise here?

We'll likely need the relative wrapper for the loading block still, but I'll make another pass to be certain.

mtias replied:
We should remove the loading block :)

aduth replied:
>We should remove the loading block :)

One PR at a time 😄

aduth commented:
>It works well for me, and I wouldn't mind merging before #4824 if that may take longer.

I want to do some last-minute performance debugging, as I was noticing an input lag in my testing. I think it may be that both the `value` assignment of the input field and the input's own `change` event (bound by `autosize`) are duplicating effort in resizing the field.

aduth commented:
This was not previously accounting for newlines, added either by enter keypress or pasted text including multiple lines. Both cases are now accounted for in 99d89b1. Flipping this back to Needs Review.

nylen commented on client/post-editor/editor-title/index.jsx:
```diff
@@ -45,37 +41,40 @@ const EditorTitle = React.createClass( {
-			} );
+		const { isNew, site } = this.props;
+		if ( ( isNew && ! prevProps.isNew ) ||
+				( isNew && get( prevProps.site, 'ID' ) !== get( site, 'ID' ) ) ) {
```
I think this indentation style is clearer here:

```js
if (
    ( isNew && ! prevProps.isNew ) ||
    ( isNew && get( prevProps.site, 'ID' ) !== get( site, 'ID' ) )
) {
    // ...
}
```

aduth replied:
>I think this indentation style is clearer here:

I go back and forth myself, but FWIW this is the documented style:

https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#multi-line-statements

nylen commented on client/post-editor/editor-title/index.jsx:
```diff
@@ -85,27 +84,48 @@ const EditorTitle = React.createClass( {
+		const newTitle = valueChars.join( '' );
+
+		this.setTitle( newTitle );
+		this.selectionStart = selectionStart + text.length;
```
Could use an explanatory comment:

```js
// Cursor position may need to change, but we have to wait for the value to update first
```

nylen commented:
Works on Chrome, Safari, Firefox, Edge.  Code looks good.

There are a couple of minor issues on IE11:

![long-title-ie11](https://cloud.githubusercontent.com/assets/227022/14690127/31381dac-0710-11e6-83f6-7e5ab857c441.gif)

- Scrollbar (apparently [can be eliminated with `overflow: auto`](https://css-tricks.com/snippets/css/remove-scrollbar-from-textarea-in-ie/))
- Cursor resets to end of line on paste.  (Attempting to debug this crashed IE in my VM.)

On iOS, I'm not able to paste because the text toolbar disappears immediately (also an issue in master - I will make a new issue), but other than that, the multi-line behavior seems fine.

aduth commented:
@nylen : Thanks for the review. Both issues should be fixed now. Neither solutions are pretty 😄  (692ecda, cab2d16).

I'm still not entirely happy with the performance of the input with these changes. I've confirmed that the update calculations aren't being performed multiple times, but it still doesn't feel quite right. I think it's likely a combination of both the height calculation and the title saving its changes directly to the post store on each change. I'd like to move away from the latter case, but I think there are UX implications of having the option to "Save" changes as soon as characters are entered into the title field.

nylen commented:
👍 :shipit:

lancewillett commented:
OK to merge? Waiting on anything? (5 days ago.)

aduth commented:
@lancewillett Works but not very happy with performance. Would be curious to get more feedback on whether it's noticeable, or at least spend more time profiling / improving performance before merging (as mentioned above, may require parallel changes to how we're tracking post title state).

nylen commented:
fwiw, I didn't notice any performance problems specific to this change.  The editor was pretty laggy in my IE11 VM, but that was a general issue and also true in master.

lancewillett commented:
I'm marking as "In Progress" so it doesn't look like it's ready to merge by anyone.

aduth commented:
Updated original comment to reflect that this incidentally fixes #1679.

lancewillett commented:
Hi @aduth — checking in on this one since last update was about a month ago. Still in progress? Can we help test a branch or change to nudge it ahead?

aduth commented:
@lancewillett : I'm still interested in this one. I proposed #5445 in hopes of potentially addressing performance concerns, but may still want to merge this one beforehand nonetheless. At the least, I'll plan to rebase and do a last round of testing, as some of the issues I may have been experiencing could have been exclusive to my IE11 VM (where issues exist in master as well).

aduth commented:
Rebased to resolve merge conflicts. @lancewillett @mtias @nylen one of you want to take this for a final spin? Else, since it'd previously been marked as ready to merge, I may move forward with merging in the next day.

I tried it again in IE11 and, even with my slow VM, it wasn't noticeable enough an issue that I'd consider performance a blocker for merging (may have improved since merging #5438).

nylen commented:
Upon retesting, I'm seeing some issues with this.  I think they are a mix of existing issues that I missed before (pasting, undo stack) and new ones (Safari resize behavior).  In the recordings below, I'm typing, then cut/pasting some text, and then hitting undo/redo.

## Chrome

In Chrome (v50), pasting should not trim the initial space, and pasting also makes undo/redo act in unexpected ways:

![editor-multiline-titles-chrome](https://cloud.githubusercontent.com/assets/227022/15874272/ff27adea-2cc7-11e6-8143-e92642670e5b.gif)

## Safari

In Safari (v9.1.1), there are multiple issues:

- pasting trims the inital space like in Chrome
- the undo stack is differently broken (only the paste event is retained, which I guess is not as bad)
- the elastic scrolling effect briefly interferes with the resizing
- the textarea does not decrease its height when it should:

![editor-multiline-titles-safari](https://cloud.githubusercontent.com/assets/227022/15874391/89b3465e-2cc8-11e6-9e4f-d9bcebb5a532.gif)

## Firefox

In Firefox, everything works really well.

aduth commented:
Thanks @nylen , these are some good catches. With regards to the pasting oddities, I feel that these may be vendor-specific behaviors? I'm not sure what the "correct" behavior is to can aim for. Do you think these will be disruptive to user's expectations?

The trimming and Safari resizing were thankfully more easy to address. I rebased and pushed 9c5219763c057e9b332564f63a7ad6d4341ca9a1 and ff5d8d4e7d2a9f01922bdc8fb357fd1f44069708 to respectively resolve the issues.

nylen commented:
Paste behavior and Safari resize behavior look good now.  I couldn't think of an easy way to address the undo stack issues either, so I'll file a separate issue for that.

Also noting that when this change is merged it becomes more important to address the remaining issues with toolbar pinning (#4907) because this adds another situation where the height of editor elements above the toolbar can change.

codebykat commented:
@aduth Did this fall through the cracks? Looks like it's been marked as ready to merge for awhile and the blocking PR is now resolved.

aduth commented:
@codebykat : Sorta. I had planned to coordinate this merging with #5445, the other being quite a bit more risky. It doesn't need to, though might benefit from some of the changes there. In any case, I can plan to revive and merge this one this week (looks to not have any merge conflicts surprisingly, but definitely warrants a rebase all the same).

codebykat commented:
Either way! I was just going through my list and it looked abandoned.

aduth commented:
Pushed a rebase including a much simpler technique for blocking newlines from pasted text. Because the title is a [controlled input](https://facebook.github.io/react/docs/forms.html#controlled-components), we can replace the newlines in the `onChange` handler, and this will be reflected both in the post edit state and in the next render pass.

See 1213941 for detail. I kept the `preventEnterKeyPress` behavior because while the `onChange` handler would prevent the newlines from enter keypresses, it'd either add a space with the current implementation, or even if we replaced newlines with an empty character, the autosize would already have updated the height in expectation that the newline had been added.

gwwar commented on client/post-editor/editor-title/style.scss:
```diff
@@ -36,14 +32,17 @@
 	color: $gray-dark;
 	font-weight: 600;
+	resize: none;
+	-ms-overflow-y: hidden !important;
```
What happens in IE when this rule isn't here?

nylen replied:
An ugly Windows native scrollbar, IIRC.

aduth replied:
>An ugly Windows native scrollbar, IIRC.

Indeed, this one: https://github.com/Automattic/wp-calypso/pull/4826#issuecomment-212604535

gwwar commented:
Found a minor corner case:

![double-paste](https://cloud.githubusercontent.com/assets/1270189/16703391/d8da9838-4521-11e6-9998-2c323679b7bc.gif)

1. Copy text into clipboard with new lines:
```text
This title
Has
Three Lines
```
2. Paste into title
3. Select Text by double clicking
4. Paste Again

gwwar commented:
Clicking into the main editor does not deselect any selected title text

gwwar commented:
I left a few notes on minor corner cases, but this feels pretty solid to me. 👍

aduth commented:
>Found a minor corner case:

Ah, this looks to be the same underlying cause as [the issue with Enter I described above](https://github.com/Automattic/wp-calypso/pull/4826#issuecomment-230885559).

>I kept the preventEnterKeyPress behavior because while the onChange replace would prevent the newlines from enter keypresses, it'd either add a space with the current implementation, or even if we replaced newlines with an empty character, the autosize would already have updated the height in expectation that the newline had been added.

I'll plan to poke a bit to see if it's straight-forward to track when the input value could change in this way and whether adding a manual height recalculation could help (and to remove the `onKeyPress` callback).

aduth commented:
I came up with an alternate solution in 66ad418 to correct the autosize height after replacing if value would have stayed the same. This also removes the need for the `onKeyPress` enter key handling.

Also rebased and squashed a bit, bumped version of `autosize` to latest patch version ([changelog](https://github.com/jackmoore/autosize/blob/master/changelog.md)), and did another test run through all major browsers.
